### ISSUE TITLE
[Weasyl] Image hosting changes cause plugin to mis-identify author and filename

### ISSUE THREAD

**Issue by Eupeptic1** [Simon-Tesla/RaccoonyWebEx#48]
The quick version is that Weasyl recently changed to use Cloudfront for their image hosting.  Since then, Raccoony is confused about what the author and filename are, although downloading with Raccoony still works correctly.

## Long version:

Around 8 or 9 Feb 2020, Weasyl put a new version of their site in production.  The new version uses Cloudfront to host or cache the image submissions.  Since then, the Weasyl plugin is having trouble parsing the author name and the filename from the page.  The plugin can still correctly find the description, title, and other metadata items, and downloading the submission in Raccoony still works, but the directory and filename the submission is saved to don't make sense.

This is with Raccoony 1.2.2 and Firefox Quantum 68.4.1esr, 64-bit, on Linux desktop.

## Steps to reproduce:

1.  Open a submission page on Weasyl in your browser.  Make a note of the artist and title of the submission.

2.  (Optional) Mouse over the raccoon-head icon, click the gear, turn "download metadata file" on, and save changes.

3.  Mouse over the raccoon-head icon, and click "Download".

4.  In your OS file manager, navigate to where Raccoony saves files (probably `~/Downloads/raccoony/weasyl` ).

5.  Look for a subdirectory with the name of that artist.

Expected result: Subdirectory of that name exists.

Actual result: Subdirectory exists, but has a hash-type name and a '.jpg' extension, something like `
Ml4HiogpR1clKVv4tcC6Cg.jpg`.

6.  In your OS file manager, go into that subdirectory, and look at the filename inside it.

Expected result:  Filename resembling `1234567_my-nifty-artwork_by_coolartist1.jpg`.

Actual result: Hash-type name with a double file extension, resembling `_Ml4HiogpR1clKVv4tcC6Cg_by_Ml4HiogpR1clKVv4tcC6Cg.jpg.jpg` .

7.  (Optional) Open the metadata text file that is in that directory, and look at the `Author:` entry in the human-readable part.

Expected result: Author/artist's name in the file matches what is displayed on Weasyl.

Actual result: Hash-type name appears instead, something like `Ml4HiogpR1clKVv4tcC6Cg.jpg`.

## Other notes

Weasyl seems to be in a state of flux right now, so it may not make sense to do any work on this until it becomes clear that they will keep using Cloudfront (or some other image-hosting/caching provider), instead of whatever they were using before.

**Comment by Eupeptic1** [comment]
> Weasyl seems to be in a state of flux right now,

The Weasyl dev chat backs this up; they may keep using S3/Cloudfront, or move to B2/Cloudflare, depending on cost.  They _may_ also go back to hosting the images on Weasyl itself, in which case Raccoony should work correctly again.

If the Weasyl plugin is modified to *only* look at the metadata it can parse from the HTML, rather than parsing the image file name, it might be more resistant to image-hosting changes at Weasyl.

Also, Weasyl has an API; documentation at https://projects.weasyl.com/weasylapi/index.html .  If Raccoony could scrape just the submission ID from the HTML, it could then get (most of?) what it needs from the API, as JSON.

However, I don't know if it's 1) possible or 2) a good idea for Raccoony to make an additional HTTP request while it's parsing the page.

**Comment by Simon-Tesla** [comment]
Yeah, I suspect that it'd make more sense to move away from parsing out filenames and move towards finding OpenGraph tags and DOM elements to fill more of that data. I've been a bit on the fence about using APIs instead of scraping for specifically the reason you describe; it adds an HTTP request and most of the time it requires a separate login that I'd have to expose UI for.

**Comment by Eupeptic1** [comment]
I have a proof-of-concept fix.  It turns out Weasyl provides a "canonical" link in the HTML head, and a lot of what we need is in that, in almost the same way it used to be in the link to the image file.

Problems:  It doesn't set serviceFilename correctly.  It might not be setting some other members of the media struct correctly.  It also doesn't parse the tags, but the current released version doesn't either (see #49).  

Tested: Against one .png and one .jpg at Weasyl; it downloaded both of them with reasonable filenames.  Also, both metadata files had reasonable values for author and filename in them.

Here it is: https://gist.github.com/Eupeptic1/95816a90d81daa919fbb58075bb82578#file-weasyl-ts

**Comment by Simon-Tesla** [comment]
Ah, awesome. And serviceFilename has always been kind of a best-effort thing, so it's not a biggie.

**Comment by Simon-Tesla** [comment]
(Work is finally not being so stressful, so I'm hoping to take advantage of the long weekend to do a bunch of bugfixing and push out a new release. Definitely appreciate the help btw n.n)

**Comment by Eupeptic1** [comment]
> Here it is: https://gist.github.com/Eupeptic1/95816a90d81daa919fbb58075bb82578#file-weasyl-ts

## The above code has errors.  It doesn't handle downloading audio files very well.

It screws up in three ways:

1.  Sometimes you get the image the artist uploaded with the audio file, and not the audio itself.  Example: https://www.weasyl.com/~kavaeric/submissions/1873551/seven-hundred-years-is-a-long-time

2.  Sometimes you don't get the raccoon-head icon at all.  This probably happens when there isn't an image with the audio file.  Example: https://www.weasyl.com/~xenosmilus/submissions/1871231/diamond-roses

3.  Sometimes you get the audio file saved to your local disk, but it has a `.jpg` extension, rather than `.mp3` or something more appropriate for audio.  Example: https://www.weasyl.com/%7Exenosmilus/submissions/1843824/kingos-the-hill-hyped-version

I discovered 1 while dog-fooding, and further investigation found 2 and 3.  The plugin still seems to handle the metadata parsing and (if selected) the metadata file OK; it just can't deal with the main audio file.

The released version of 1.2.2 does the right thing for all three examples above.

I plan to work on this in the next day or two.

**Comment by Eupeptic1** [comment]
`// Get ext from previewUrl
// FIXME why not get it from the URL from the download button?`

This is why.  :D  Fix is prototyped and works on the three cases above, but I want to test it some more.

**Comment by Eupeptic1** [comment]
Revised version is here: https://gist.github.com/Eupeptic1/df8355074ba9f0772ac83b1041a576b7

I changed it to get the extension (only) from the link in the native download button.  It still gets most other things from the canonical link.

I also discovered that the canonical link provided by Weasyl has at least two slightly different formats, not one, so I added code to handle the two formats I think we care about.

I tried to set serviceFilename to something hopefully reasonable.

I included the fix for the tags not showing up in the metadata file, as discussed in #49 .

Problems: It still has a lot of logging statements.

Tested for individual submissions: Against regular images, regular music (multimedia), regular story, and characters.

Tested for "open all in tabs": the main home page, the "Recently Popular" page, and my own profile page.
